### DISCUSSION TITLE
fatal: Not a valid object name HEAD

### DISCUSSION THREAD

**Discussion by xiaonig** [neuralhydrology/neuralhydrology#177] [Q&A] [answered]
Hello. I create the environment
```
conda env create -f environment_cuda11_8.yml
conda activate neuralhydrology
pip install neuralhydrology
```
When running introduction.ipynb second cell, it reports an error.
```
fatal: Not a valid object name HEAD
CalledProcessError: Command '['git', '-C', '/home/.conda/envs/neuralhydrology/lib/python3.10/site-packages/neuralhydrology/utils', 'describe', '--always']' returned non-zero exit status 128.
```
Looking forward to your reply, thanks!

**Comment by gauchm** [comment]
Hi,
we had this issue pop up once before in #123, but we never understood the root cause. You can find a workaround there (see [this reply](https://github.com/neuralhydrology/neuralhydrology/issues/123#issuecomment-1572337130)).

But it would nevertheless be nice to understand why/how this is happening. So it would be great if you could try to run three things:
1. `git -C /home/.conda/envs/neuralhydrology/lib/python3.10/site-packages/neuralhydrology/utils rev-parse --show-toplevel`
2. `git -C /home/.conda/envs/neuralhydrology/lib/python3.10/site-packages/neuralhydrology/utils describe --always`
3. `git -C /home/.conda/envs/neuralhydrology/lib/python3.10/site-packages/neuralhydrology/utils branch` 

and report the output of these commands (please try to do this before you apply the workaround).
Thanks for helping!

**Comment by xiaonig** [comment]
Thanks for the reply.Trying to change the previous operation results in:
```
(neuralhydrology) mengziwen@vm-jupyterhub-server:~/neuralhydrology/environments$ git -C /home/mengziwen/.conda/envs/neuralhydrology/lib/python3.10/site-packages/neuralhydrology/utils rev-parse --show-toplevel
/home/mengziwen
(neuralhydrology) mengziwen@vm-jupyterhub-server:~/neuralhydrology/environments$ git -C /home/mengziwen/.conda/envs/neuralhydrology/lib/python3.10/site-packages/neuralhydrology/utils describe --always
fatal: Not a valid object name HEAD
(neuralhydrology) mengziwen@vm-jupyterhub-server:~/neuralhydrology/environments$ git -C /home/mengziwen/.conda/envs/neuralhydrology/lib/python3.10/site-packages/neuralhydrology/utils branch
(neuralhydrology) mengziwen@vm-jupyterhub-server:~/neuralhydrology/environments$
```
But it still doesn't work when I try new methods. I tried both accounts in the cloud jupyter hub provided by the team and all of them were unsuccessful, but when I installed the environment locally, it worked. Maybe there's something wrong with the platform on my side, I'll go to the team's techs in a while and ask if there's a problem with the platform

**Reply by gauchm** [reply, marked as answer]
The output of the first command looks like your home directory (`/home/mengziwen`) is a git repo... is this intentional? Since there is no HEAD, I believe you have never made any commit in that git repo, so my guess is it's not intentional.

If it's not intentional, probably deleting the `.git` folder that should be present in ` /home/mengziwen` would help (CAREFUL: this will delete the history of that git repo. If I'm correct and you never used that repo this is fine, but if you're unsure about this, maybe just rename the `.git` folder to `.tbd.git` or something else so you don't lose it).

For background on what's happening: neuralhydrology checks whether its code operates in a git repo, and if it does, it tries to get the commit hash of that repo for logging purposes. Now in your case, the code does operate in a git repo, but your git repo has never had any commits, so there is no commit hash to get. If you want to keep your git repo for some reason, you could also solve the error by making a commit in there.

**Reply by xiaonig** [reply]
Thank you for your response. We just finished a six-day long business trip. I just deleted the `.git` and then it doesn't report errors when I run it again. I have a general understanding of how neuralhydrology works and appreciate your guidance!
